rest/api.go

```diff
@@ -278,32 +280,39 @@ func (h *handler) handleProfiling() error {
 		}
 	}
 
-	if params.File != "" {
-		f, err := os.Create(params.File)
-		if err != nil {
-			return err
+	// Handle no file
+	if params.File == "" {
+		if isCPUProfile {
+			base.InfofCtx(h.db.Ctx, base.KeyAll, "... ending CPU profile")
+			pprof.StopCPUProfile()
+			return nil
 		}
-		if profileName != "" {
-			defer f.Close()
-			if profile := pprof.Lookup(profileName); profile != nil {
-				profile.WriteTo(f, 0)
-				base.Infof(base.KeyAll, "Wrote %s profile to %s", profileName, base.UD(params.File))
-			} else {
-				return base.HTTPErrorf(http.StatusNotFound, "No such profile %q", profileName)
-			}
-		} else {
-			base.Infof(base.KeyAll, "Starting CPU profile to %s ...", base.UD(params.File))
-			pprof.StartCPUProfile(f)
+		return base.HTTPErrorf(http.StatusBadRequest, "Missing JSON 'file' parameter")
+	}
+
+	f, err := os.Create(params.File)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		fErr := f.Close()
```

Comment:
We shouldn't be closing the file for StartCPUProfile or StopCPUProfile calls - that's managed by pprof.StopCPUProfile.  Should move this out of the defer down to line 314,  and return early for the StartCPUProfile case.

Reply:
Changes are committed to the latest revision.